tls/s2n_renegotiate.c

```diff
@@ -0,0 +1,125 @@
+ */
+
+#include "tls/s2n_renegotiate.h"
+
+#include "error/s2n_errno.h"
+#include "stuffer/s2n_stuffer.h"
+#include "tls/s2n_connection.h"
+#include "utils/s2n_safety.h"
+
+/*
+ * Prepare a connection to be reused for a second handshake.
+ *
+ * s2n-tls was not originally designed to support renegotiation.
+ * s2n_connection is a very large structure with configuration fields set by the application
+ * mixed in with internal state fields set by the handshake.
+ * Ensuring all existing internal state fields (and any new fields added) are safe to reuse for
+ * a renegotiated handshake would be extremely prone to errors.
+ * For safety, we instead wipe the entire connection and only restore fields we know we need
+ * in order to continue sending and receiving encrypted data.
+ *
+ * Any configuration fields set by the application will need to be set by the application again.
+ */
+int s2n_renegotiate_wipe(struct s2n_connection *conn)
+{
+    POSIX_ENSURE_REF(conn);
+
+    /* We use this method to reset both clients and servers when testing.
+     * However, outside of tests, it should only be called for client connections
+     * because we only support renegotiation for clients.
+     */
+    POSIX_ENSURE(conn->mode == S2N_CLIENT || s2n_in_unit_test(), S2N_ERR_NO_RENEGOTIATION);
+
+    /* Best effort check for pending input or output data.
+     * This method should not be called until the application has stopped sending and receiving.
+     * Saving partial read or parital write state would complicate this problem.
+     */
+    POSIX_ENSURE(s2n_stuffer_data_available(&conn->header_in) == 0, S2N_ERR_INVALID_STATE);
+    POSIX_ENSURE(s2n_stuffer_data_available(&conn->in) == 0, S2N_ERR_INVALID_STATE);
+    POSIX_ENSURE(s2n_stuffer_data_available(&conn->out) == 0, S2N_ERR_INVALID_STATE);
+
+    /* Save the crypto parameters.
+     * We need to continue encrypting / decrypting with the old secure parameters.
+     */
+    DEFER_CLEANUP(struct s2n_crypto_parameters *secure_crypto_params = conn->secure,
+            s2n_crypto_parameters_free);
+    conn->secure = NULL;
+
+    /* Save the fragment length so we continue properly fragmenting our records
+     * until a new fragment length is chosen.
+     */
+    uint16_t max_frag_len = conn->max_outgoing_fragment_length;
+
+    /* Save the protocol versions.
+     * Various checks when sending and receiving records rely on the protocol version. */
+    uint8_t actual_protocol_version = conn->actual_protocol_version;
+    uint8_t server_protocol_version = conn->server_protocol_version;
+    uint8_t client_protocol_version = conn->client_protocol_version;
+    POSIX_ENSURE(actual_protocol_version < S2N_TLS13, S2N_ERR_PROTOCOL_VERSION_UNSUPPORTED);
+
+    /* Save byte tracking.
+     * This isn't strictly necessary, but potentially useful. */
+    uint64_t wire_bytes_in = conn->wire_bytes_in;
+    uint64_t wire_bytes_out = conn->wire_bytes_out;
+
+    /* Save io settings */
```

Comment:
The send/receive callbacks can be set by the user, right? Why are these callbacks preserved while other callbacks are not? Wouldn't the user expect to set these again with all the other stuff they set after wiping?

Reply:
The general idea is that I want IO to continue to work after the wipe. If we don't have any IO methods, we can't do IO.

Depending on how the user handles IO, it might also be fairly tricky to "re-setup" IO. We don't want to reset or close or whatever the underlying socket.

However, if the user does manage to "re-setup" IO without breaking the connection, it shouldn't hurt anything. It'd just be unnecessary.

Reply:
Got it, okay. That makes sense. Thanks!